### PR TITLE
[Core] Brep Surface - trimmed integration and clipper lib

### PR DESCRIPTION
**Pull request by tteschemacher**
hey all,

as there has been some changes within the implementation of the integration info and the according interfaces I recreated a new PR for the creation of the integration of a trimmed domain. Accordingly it deprecates #8325 .

This procedure unfortunately requires an external library, the clipper lib.

### CODE REVIEW

File: kratos/utilities/geometry_utilities/brep_trimming_utilities.h [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+//                   Kratos default license: kratos/license.txt
+
+#if !defined(KRATOS_BREP_TRIMMING_UTILITIES_H_INCLUDED)
+#define KRATOS_BREP_TRIMMING_UTILITIES_H_INCLUDED
+
+// System includes
+#include "includes/define.h"
+
+// External includes
+#include "clipper/clipper.h"
```
Comment (philbucher) [inline kratos/utilities/geometry_utilities/brep_trimming_utilities.h]: I strongly suggest to only include this in a cpp, as otherwise this will be propagated to lots of places.
I see you templated the code but in this case I would pay the price of instantiating the required classes

File: kratos/CMakeLists.txt [hunk 2]
```diff
@@ -63,7 +63,7 @@ target_compile_definitions(KratosVersion PRIVATE
 
 ## Define library KratosCore to be included in all of the others
 add_library(KratosCore SHARED ${KRATOS_CORE_SOURCES} ${KRATOS_CORE_TESTING_ENGINE_SOURCES} ${KRATOS_CORE_INPUT_OUTPUT_SOURCES} ${KRATOS_TEST_SOURCES} $<TARGET_OBJECTS:KratosVersion>)
-target_link_libraries(KratosCore PUBLIC gidpost triangle)
+target_link_libraries(KratosCore PUBLIC gidpost triangle clipper)
```
Comment (roigcarlo) [inline kratos/CMakeLists.txt]: Is there really no other alternative than compiling this and add it as a dependency in the linking stage?

Comment (tteschemacher) [reply kratos/CMakeLists.txt]: I'd be open for any suggestion?

**Comment by RiccardoRossi** [comment]
@KratosMultiphysics/technical-committee agrees in principle about this addition. however we will take a deeper look into the details of the implementation

In particular, why is it needed to add integration new integration rules outside of the triangle geometry? cannot we add them in there?

File: external_libraries/clipper/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+# generate variables with the sources
+set(CLIPPER_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/clipper.cpp)
+
+add_library(clipper STATIC ${CLIPPER_SOURCES})
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+  target_link_libraries(clipper m)
+endif(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
```
Comment (philbucher) [inline external_libraries/clipper/CMakeLists.txt]: why is this necessary?
Does it works without?

File: external_libraries/clipper/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+set(CLIPPER_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+set(CLIPPER_LIBRARIES clipper)
+
+message(STATUS "clipper : ${CLIPPER_INCLUDE_DIR}" )
+# install(TARGETS triangle DESTINATION libs )
```
Comment (philbucher) [inline external_libraries/clipper/CMakeLists.txt]:
```suggestion
```

**Review by philbucher — COMMENTED** [review]
in terms of includes the library seems ok (no `windows.h`)

File: external_libraries/clipper/CMakeLists.txt (hunk 3 above)
Comment (tteschemacher) [reply external_libraries/clipper/CMakeLists.txt]: That was a copy and paste from triangle. I have removed it for the moment to check if it works.

**Comment by tteschemacher** [comment]
> @KratosMultiphysics/technical-committee agrees in principle about this addition. however we will take a deeper look into the details of the implementation


Many thanks for checking this PR! Highly appreciated.

> In particular, why is it needed to add integration new integration rules outside of the triangle geometry? cannot we add them in there?

First of all, here, a much higher order is being considered. Adding this to the triangle would somehow decrease the flexibility of the triangles as there it would be stored once for each derivative.

Secondly, this is the much more efficient access to the integration of the location points. Otherwise, either triangles would need to be created and the access would need to happen through the IntegrationMethod, which generally is very restrained. With this type of access not even a single copy of integration point locations will be required.

The same we already do for the 1-dimensional gauss points.

**Comment by RiccardoRossi** [comment]
well..the license is compatible, and we want this capability.

i would allow this in...

**Review by pooyan-dadvand — APPROVED** [review]
@KratosMultiphysics/technical-committee approves it as according to http://www.angusj.com/clipper2/Docs/License.htm it is compatible with our license and needed

**Comment by loumalouomega** [comment]
Interesting, I may check this library, maybe I can replace the manual clipping I use in mortar and even improve performance

**Comment by tteschemacher** [comment]
Yay.
